src/OrchardCore/OrchardCore.Mvc.Core/LocationExpander/SharedViewLocationExpanderProvider.cs

```diff
@@ -0,0 +1,142 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Mvc.Razor;
+using Microsoft.Extensions.Caching.Memory;
+using OrchardCore.Environment.Extensions;
+using OrchardCore.Environment.Shell.Descriptor.Models;
+using OrchardCore.Mvc.FileProviders;
+
+namespace OrchardCore.Mvc.LocationExpander
+{
+    public class SharedViewLocationExpanderProvider : IViewLocationExpanderProvider
+    {
+        private const string CacheKey = "ModuleSharedViewLocations";
+        private const string PageCacheKey = "ModulePageSharedViewLocations";
+        private static IList<IExtensionInfo> _modulesWithPageSharedViews;
+        private static IList<IExtensionInfo> _modulesWithSharedViews;
+        private static object _synLock = new object();
+
+        private readonly IExtensionManager _extensionManager;
+        private readonly ShellDescriptor _shellDescriptor;
+        private readonly IMemoryCache _memoryCache;
+
+        public SharedViewLocationExpanderProvider(
+            RazorCompilationFileProviderAccessor fileProviderAccessor,
+            IExtensionManager extensionManager,
+            ShellDescriptor shellDescriptor,
+            IMemoryCache memoryCache)
+        {
+            _extensionManager = extensionManager;
+            _shellDescriptor = shellDescriptor;
+            _memoryCache = memoryCache;
+
+            if (_modulesWithSharedViews != null)
+            {
+                return;
+            }
+
+            lock (_synLock)
+            {
+                if (_modulesWithSharedViews == null)
+                {
+                    var orderedModules = _extensionManager.GetExtensions()
+                        .Where(e => e.Manifest.Type.Equals("module", StringComparison.OrdinalIgnoreCase))
+                        .Reverse();
+
+                    var modulesWithPageSharedViews = new List<IExtensionInfo>();
+                    foreach (var module in orderedModules)
```

Comment:
I think we can do both "Pages/Shared" 7 "Views/Shared" in the same loop

Reply:
Yes, you're right, thanks. Just added the `Page` part too quickly